File: trunk/C#/Sample Project/SampleProject/SampleProject.Exceptions/CustomExceptions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text; 
using System.Web;
using Microsoft.SharePoint;
using Microsoft.SharePoint.Administration;
using SampleProject.Common;

namespace SampleProject.Exceptions
{
    public class CustomExceptions : Exception
    {
        private Exception innerException;

        public CustomExceptions(Exception exception, Enums.ShowException showException)
        {
            this.innerException = exception;
            switch (showException)
            {
                case Enums.ShowException.EventViewer:
                    ShowMessageonEventViewer();
                    break;
                case Enums.ShowException.LogFile:
                    ShowMessageonLogFile();
                    break;
                default:
                    break;
            }
        }

        public CustomExceptions(Exception exception, HttpContext current, SPWeb web)
        {
            this.innerException = exception;
            ShowMessageonUI(current, web);
        }

        #region Private Methods
        private void ShowMessageonEventViewer()
        {
            System.Diagnostics.EventLog.WriteEntry(Constants.ERROR_KEY , this.innerException.Message + ": " + this.innerException.StackTrace, System.Diagnostics.EventLogEntryType.Error);
        }

        private void ShowMessageonLogFile()
        {
            SPDiagnosticsService.Local.WriteTrace(0, new SPDiagnosticsCategory(Constants.ERROR_KEY, TraceSeverity.Medium, EventSeverity.Error), TraceSeverity.Monitorable, this.innerException.Message, this.innerException.StackTrace);
        }

        private void ShowMessageonUI(HttpContext current, SPWeb web)
        {
            SetErrorMessageforUI(current, web, this.innerException.Message);
        }
        #endregion
      
        #region UI Message
       
        #region Public Clear Message

        public static void ClearMessageFromUI(HttpContext current, SPWeb web)
        {
            if (current != null)
            {
                SetMessageVariable(current, web, String.Empty);
            }
        }
        #endregion

        #region Public Set Error Message
        public static void SetErrorMessageforUI(HttpContext current, SPWeb web, String sErrorMessage)
        {
            if (current != null)
            {
                if (!String.IsNullOrEmpty(sErrorMessage))
                {
                    String sMessage = Constants.ERROR_MESSAGE_SYMBOL + sErrorMessage;
                    SetMessageVariable(current, web, sMessage);
                }
            }
        }
        #endregion

        #region Public Set Success Message
        public static void SetSuccessMessageforUI(HttpContext current, SPWeb web, String sErrorMessage)
        {
            if (current != null)
            {
                if (!String.IsNullOrEmpty(sErrorMessage))
                {
                    String sMessage = Constants.SUCCESS_MESSAGE_SYMBOL + sErrorMessage;
                    SetMessageVariable(current, web, sMessage);
                }
            }
        }

        #endregion

        public static void SetMessageVariable(HttpContext current, SPWeb web, String sErrorMessage)
        {
            if (current != null)
            {
                current.Cache[CommonFunctions.GetErrorCodeKey(current, web)] = sErrorMessage;
            }
        }
        #endregion
    }
}

File: trunk/VB/Sample Project/SampleProject/SampleProject.Exceptions/CustomExceptions.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Web
Imports Microsoft.SharePoint
Imports Microsoft.SharePoint.Administration
Imports SampleProject.Common

Public Class CustomExceptions
    Inherits Exception
    Private innerExceptionNew As Exception

    Public Sub New(ByVal exception As Exception, ByVal showException As Enums.ShowException)
        Me.innerExceptionNew = exception
        Select Case showException
            Case Enums.ShowException.EventViewer
                ShowMessageonEventViewer()
                Exit Select
            Case Enums.ShowException.LogFile
                ShowMessageonLogFile()
                Exit Select
            Case Else
                Exit Select
        End Select
    End Sub

    Public Sub New(ByVal exception As Exception, ByVal current As HttpContext, ByVal web As SPWeb)
        Me.innerExceptionNew = exception
        ShowMessageonUI(current, web)
    End Sub

#Region "Private Methods"
    Private Sub ShowMessageonEventViewer()
        System.Diagnostics.EventLog.WriteEntry(Constants.ERROR_KEY, Me.innerExceptionNew.Message & ": " & Me.innerExceptionNew.StackTrace, System.Diagnostics.EventLogEntryType.[Error])
    End Sub

    Private Sub ShowMessageonLogFile()
        SPDiagnosticsService.Local.WriteTrace(0, New SPDiagnosticsCategory(Constants.ERROR_KEY, TraceSeverity.Medium, EventSeverity.[Error]), TraceSeverity.Monitorable, Me.innerExceptionNew.Message, Me.innerExceptionNew.StackTrace)
    End Sub

    Private Sub ShowMessageonUI(ByVal current As HttpContext, ByVal web As SPWeb)
        SetErrorMessageforUI(current, web, Me.innerExceptionNew.Message)
    End Sub
#End Region

#Region "UI Message"

#Region "Public Clear Message"

    Public Shared Sub ClearMessageFromUI(ByVal current As HttpContext, ByVal web As SPWeb)
        If current IsNot Nothing Then
            SetMessageVariable(current, web, [String].Empty)
        End If
    End Sub
#End Region

#Region "Public Set Error Message"
    Public Shared Sub SetErrorMessageforUI(ByVal current As HttpContext, ByVal web As SPWeb, ByVal sErrorMessage As [String])
        If current IsNot Nothing Then
            If Not [String].IsNullOrEmpty(sErrorMessage) Then
                Dim sMessage As [String] = Constants.ERROR_MESSAGE_SYMBOL + sErrorMessage
                SetMessageVariable(current, web, sMessage)
            End If
        End If
    End Sub
#End Region

#Region "Public Set Success Message"
    Public Shared Sub SetSuccessMessageforUI(ByVal current As HttpContext, ByVal web As SPWeb, ByVal sErrorMessage As [String])
        If current IsNot Nothing Then
            If Not [String].IsNullOrEmpty(sErrorMessage) Then
                Dim sMessage As [String] = Constants.SUCCESS_MESSAGE_SYMBOL + sErrorMessage
                SetMessageVariable(current, web, sMessage)
            End If
        End If
    End Sub

#End Region

    Public Shared Sub SetMessageVariable(ByVal current As HttpContext, ByVal web As SPWeb, ByVal sErrorMessage As [String])
        If current IsNot Nothing Then
            current.Cache(CommonFunctions.GetErrorCodeKey(current, web)) = sErrorMessage
        End If
    End Sub
#End Region
End Class
